### ISSUE TITLE
error while compiling another code

### ISSUE THREAD

**Issue by o-kazemi** [kokkos/kokkos#5704]
Hello there, 
When I try to compile a code, I get an error which arises fro kokkos side. Since the technical support of the package do not respond, I got to ask you to see if you could help me. I have pasted the error message in a text file because of the long text. Please find it here: https://github.com/o-kazemi/shared/blob/main/error 
Thanks in advance for your help.

**Comment by JBludau** [comment]
This looks like it is somehow not compiling with `std=c++14` or above. Could you check that by looking at the compile-line (e.g. using `make VERBOSE=1`)?

**Comment by masterleinad** [comment]
It would also be helpful to know which `Kokkos` release you are using, how you configured `Kokkos`, and how you link with `Kokkos` (`CMake`, `Makefiles`,...)?

**Comment by dalg24** [comment]
As Jakob pointed out, whatever you are building has not been configured properly (does not support C++14).

There is not enough information to help much beyond that (what version of Trilinos are you using, how did you configure it, how is the code you are building configured to link against Kokkos, etc.) and here is not the right place to report this kind of issue.

**Comment by o-kazemi** [comment]
The packages I have installed are:
PETSC, PetIGA, Trilion (version 13.4.1, configured with CMake), igakit and the mechanoChemIGA on the top on them. I have not configured the kokkos myself. Honestly, there was no heads up about kokkos being used! When I try to compile the examples of mechanoChemIGA using Cmake for the Cmakeslist to create a make file and then make for running the code, I get those errors.  How can I resolve this issue?

**Comment by dalg24** [comment]
Try configuring with `-DCMAKE_CXX_STANDARD=14`

**Comment by crtrott** [comment]
Generally, it looks like you do not properly inherit the required CXX standard, and or are not using a compiler which has C++14. For example you may compile using Clang but pointing to system header files for the C++ standard library. Hard to tell without more details. All in all this is not a Kokkos bug, but some downstream configure  and or system issue. That said: what Damien just posted (i.e. trying to configure the final thing with -DCMAKE_CXX_STANDARD=14) may work.

**Comment by crtrott** [comment]
Oh and generally this should not happen if everyone in the stack uses CMake properly. But some projects write completely self build build systems, and then don't end up setting the right flags, and thus can't build.

**Comment by o-kazemi** [comment]
Thanks for comprehensive reply. What other information you need to find out the problem's root? Sorry I am so illiterate. I tried to recompile the trilions package, and pasted the output here: https://github.com/o-kazemi/shared/blob/main/trilions%20installation
Does it help to detect the problem?

**Comment by dalg24** [comment]
Presumably you don't get an error when you build and then install Trilinos, you get it downstream.
Make sure you pass `-DCMAKE_CXX_STANDARD=14` when you configure that code downstream.

**Comment by o-kazemi** [comment]
No I don't! The Trilions installation process goes ahead smoothly. I tried
to include the command you mentioned above in the Cmake call of
Cmakeslists.txt, and then try to compile with make command. It seems that
it helped. But the problem is still inconsistency between various packages!
The error I get now is like below:
(base) ***@***.***:~/Program/mechanoChemIGA-master/initBounValProbs/CahnHilliard_oneSpecies/2D$
cmake -DCMAKE_CXX_STANDARD=14 CMakeLists.txt
CMake Deprecation Warning at CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- Enabled Kokkos devices: SERIAL

Found Trilinos

-- Configuring done
-- Generating done
-- Build files have been written to:
/home/omid/Program/mechanoChemIGA-master/initBounValProbs/CahnHilliard_oneSpecies/2D
(base) ***@***.***:~/Program/mechanoChemIGA-master/initBounValProbs/CahnHilliard_oneSpecies/2D$
make
Consolidate compiler generated dependencies of target main
[  8%] Building CXX object
CMakeFiles/main.dir/home/omid/Program/mechanoChemIGA-master/src/fields.cc.o
In file included from /home/omid/Program/petsc-3.18.2/include/petscbag.h:5,
                 from /home/omid/Program/petsc-3.18.2/include/petsc.h:5,
                 from /home/omid/Program/PetIGA-master/include/petiga.h:10,
                 from
/home/omid/Program/mechanoChemIGA-master/src/fields.cc:2:
/home/omid/Program/petsc-3.18.2/include/petscsys.h:95:6: error: #error
"PETSc was configured with one MPICH mpi.h version but now appears to be
compiling using a different MPICH mpi.h version"
   95 |     #error "PETSc was configured with one MPICH mpi.h version but
now appears to be compiling using a different MPICH mpi.h version"
      |      ^~~~~
make[2]: *** [CMakeFiles/main.dir/build.make:76:
CMakeFiles/main.dir/home/omid/Program/mechanoChemIGA-master/src/fields.cc.o]
Error 1
make[1]: *** [CMakeFiles/Makefile2:86: CMakeFiles/main.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
(base) ***@***.***:~/Program/mechano

**Comment by crtrott** [comment]
I looked at the examples in that code (mechanoChemIGA) and the CMakelists in there don't properly inherit from the libraries they depend on. They also explicitly set CMAKE_CXX_FLAGS to include -std=c++11. maybe it works if you just change that to c++14. if not someone needs to fix their build system.

**Comment by o-kazemi** [comment]
You are right. Now I get the error of inconsistent mpi.h. It says error: #error "PETSc was configured with one MPICH mpi.h version but now appears to be compiling using a different MPICH mpi.h version" I don't know how check the versions and how to adjust them.

**Comment by masterleinad** [comment]
You need to make sure that you use a consistent MPI implementation for all the packages that use it. At this point, this has nothing to do with `Kokkos` anymore.
